tortoise/tortoise.go

```diff
@@ -180,10 +182,6 @@ func (t *turtle) encodeVotes(
 	start types.LayerID,
 	current types.LayerID,
 ) (*types.Opinion, error) {
-	logger := t.logger.WithContext(ctx).WithFields(
-		log.Stringer("base layer", base.layer),
-		log.Stringer("current layer", current),
-	)
```

Comment:
I think a dedicated instance of logger with pre-set fields should consume less memory than re-declaring all fields on every log, since it only has to be allocated once:

```golang
logger := t.logger.With(
	log.ZContext(ctx),
	zap.Stringer("base layer", base.layer),
	zap.Stringer("current layer", current),
)
```

And then re-using `logger` should theoretically consume less memory than repeating the fields on every log.

Reply:
no this allocates logger and a lot of fields regardless if this fields will be even used,  this should not be used like that at all

Reply:
Again I'm not sure about this:

```golang
func Benchmark_DedicatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b)

	logger := log.With(
		zap.Stringer("base layer", bytes.NewBufferString("10")),
		zap.Stringer("current layer", bytes.NewBufferString("20")),
	)

	for i := 0; i < b.N; i++ {
		logger.Info("test")
	}
}

func Benchmark_RepeatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b)

	for i := 0; i < b.N; i++ {
		log.Info("test",
			zap.Stringer("base layer", bytes.NewBufferString("10")),
			zap.Stringer("current layer", bytes.NewBufferString("20")),
		)
	}
}
```

Results:

```shell
Benchmark_DedicatedLogger-6       503180              5341 ns/op            1497 B/op         14 allocs/op
--- BENCH: Benchmark_DedicatedLogger-6
    logger.go:130: 2023-06-15T09:07:58.700Z     INFO    test    {"base layer": "10", "current layer": "20"}
        ... [output truncated]
Benchmark_RepeatedLogger-6        461164              6787 ns/op            1790 B/op         21 allocs/op
--- BENCH: Benchmark_RepeatedLogger-6
    logger.go:130: 2023-06-15T09:08:01.417Z     INFO    test    {"base layer": "10", "current layer": "20"}
        ... [output truncated]
PASS
```

Reply:
wdym?

Reply:
In both cases the fields are evaluated lazily, so even if there is transformation / encoding caused by calling `String()` this only happens when the log is actually printed?

Reply:
@fasmat, consider:
```go
func Benchmark_DedicatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b, zaptest.Level(zap.ErrorLevel))
	baseLayer := bytes.NewBufferString("10")
	currentLayer := bytes.NewBufferString("20")

	for i := 0; i < b.N; i++ {
		logger := log.With(
			zap.Stringer("base layer", baseLayer),
			zap.Stringer("current layer", currentLayer),
		)
		logger.Info("test")
	}
}

func Benchmark_RepeatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b, zaptest.Level(zap.ErrorLevel))
	baseLayer := bytes.NewBufferString("10")
	currentLayer := bytes.NewBufferString("20")

	for i := 0; i < b.N; i++ {
		log.Info("test",
			zap.Stringer("base layer", baseLayer),
			zap.Stringer("current layer", currentLayer),
		)
	}
}
```

>Benchmark_DedicatedLogger-20    	 2257372	       539.4 ns/op	    1414 B/op	       8 allocs/op
Benchmark_RepeatedLogger-20     	32398957	        35.94 ns/op	     128 B/op	       1 allocs/op

Reply:
@poszu your test is incorrect, you are allocating a new logger every time which is not the case in the real code:

```golang
func Benchmark_DedicatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b, zaptest.Level(zap.ErrorLevel))
	baseLayer := bytes.NewBufferString("10")
	currentLayer := bytes.NewBufferString("20")

	logger := log.With(
		zap.Stringer("base layer", baseLayer),
		zap.Stringer("current layer", currentLayer),
	)

	for i := 0; i < b.N; i++ {
		logger.Info("test")
	}
}

func Benchmark_RepeatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b, zaptest.Level(zap.ErrorLevel))
	baseLayer := bytes.NewBufferString("10")
	currentLayer := bytes.NewBufferString("20")

	for i := 0; i < b.N; i++ {
		log.Info("test",
			zap.Stringer("base layer", baseLayer),
			zap.Stringer("current layer", currentLayer),
		)
	}
}
```

> Benchmark_DedicatedLogger-6   	262413741	         4.553 ns/op	       0 B/op	       0 allocs/op
> Benchmark_RepeatedLogger-6    	30456820	        40.82 ns/op	     128 B/op	       1 allocs/op

Reply:
no, this is wrong. With() will allocate immediately

Reply:
It is exactly the case, that's why I moved it into the loop. The point is that (before this PR changes) the `logger` was created on each function call, regardless of the log level. In effect, we pay the cost even if the log is not rendered in the end.

Reply:
> no, this is wrong. With() will allocate immediately

It creates a copy of the logger, if the original logger had lazy evaluated fields those will still not be evaluated.

> The point is that (before this PR changes) the logger was created on each function call, regardless of the log level.

But this is still comparing apples to oranges.

The creation of a child logger costs time and memory yes, but we save that on every subsequent call to the logger, because the fields don't have to be evaluated again. So it depends on how many logs statements we do with the new instance:

```golang
func Benchmark_DedicatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b, zaptest.Level(zap.ErrorLevel))
	baseLayer := bytes.NewBufferString("10")
	currentLayer := bytes.NewBufferString("20")

	for i := 0; i < b.N; i++ {
		logger := log.With(
			zap.Stringer("base layer", baseLayer),
			zap.Stringer("current layer", currentLayer),
		)

		for i := 0; i < 15; i++ {
			logger.Info("test")
		}
	}
}

func Benchmark_RepeatedLogger(b *testing.B) {
	log := zaptest.NewLogger(b, zaptest.Level(zap.ErrorLevel))
	baseLayer := bytes.NewBufferString("10")
	currentLayer := bytes.NewBufferString("20")

	for i := 0; i < b.N; i++ {
		for i := 0; i < 15; i++ {
			log.Info("test",
				zap.Stringer("base layer", baseLayer),
				zap.Stringer("current layer", currentLayer),
			)
		}
	}
}
```

> Benchmark_DedicatedLogger-6   	 1984636	       538.2 ns/op	    1412 B/op	       8 allocs/op
> Benchmark_RepeatedLogger-6    	 2044924	       603.3 ns/op	    1920 B/op	      15 allocs/op

The breakeven is somewhere between 10 and 15 logs, where the child logger becomes more efficient than repeating fields.

Reply:
> It creates a copy of the logger, if the original logger had lazy evaluated fields those will still not be evaluated.

it will call String() method even if log is not printed

Reply:
>Try it yourself!

![image](https://github.com/spacemeshos/go-spacemesh/assets/3966964/ed4fdf6d-e483-4ee4-859e-01702f84d991)

Reply:
```

type stringer struct {
}

func (s *stringer) String() string {
	fmt.Println("tried it myself")
	return ""
}

func TestString(t *testing.T) {
	log := logtest.New(t).Zap().With(zap.Stringer("s", &stringer{}))
	log.Debug("dassdsa")
}
```

> TEST_LOG_LEVEL=info go test ./tortoise -run=TestString -v
 === RUN   TestString
tried it myself
--- PASS: TestString (0.00s)
PASS
ok  	github.com/spacemeshos/go-spacemesh/tortoise	(cached

Reply:
![](https://media0.giphy.com/media/K3RxMSrERT8iI/giphy.gif)